File: ts/instructions/createTree.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface CreateTreeArgs {
  tag: string
}

export interface CreateTreeAccounts {
  signer: PublicKey
  admin: PublicKey
  /** The account that manages tokens */
  forestAuthority: PublicKey
  /** The forest */
  forest: PublicKey
  /** Mint of the token used to pay the tree creation fee */
  voteMint: PublicKey
  creatorAccount: PublicKey
  adminAccount: PublicKey
  /** The tree */
  tree: PublicKey
  /** The root node of the new tree */
  rootNode: PublicKey
  /** Common Solana programs */
  tokenProgram: PublicKey
  associatedTokenProgram: PublicKey
  systemProgram: PublicKey
  rent: PublicKey
}

export const layout = borsh.struct([borsh.str("tag")])

export function createTree(args: CreateTreeArgs, accounts: CreateTreeAccounts) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.signer, isSigner: true, isWritable: true },
    { pubkey: accounts.admin, isSigner: false, isWritable: false },
    { pubkey: accounts.forestAuthority, isSigner: false, isWritable: false },
    { pubkey: accounts.forest, isSigner: false, isWritable: false },
    { pubkey: accounts.voteMint, isSigner: false, isWritable: false },
    { pubkey: accounts.creatorAccount, isSigner: false, isWritable: true },
    { pubkey: accounts.adminAccount, isSigner: false, isWritable: true },
    { pubkey: accounts.tree, isSigner: false, isWritable: true },
    { pubkey: accounts.rootNode, isSigner: false, isWritable: true },
    { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
    {
      pubkey: accounts.associatedTokenProgram,
      isSigner: false,
      isWritable: false,
    },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
    { pubkey: accounts.rent, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([165, 83, 136, 142, 89, 202, 47, 220])
  const buffer = Buffer.alloc(1000)
  const len = layout.encode(
    {
      tag: args.tag,
    },
    buffer
  )
  const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len)
  const ix = new TransactionInstruction({ keys, programId: PROGRAM_ID, data })
  return ix
}

File: programs/dippies_index_protocol/src/instructions/create_tree.rs
use anchor_lang::prelude::*;
use anchor_spl::associated_token::AssociatedToken;
use anchor_spl::token::{transfer, Mint, Token, TokenAccount, Transfer};

use crate::constants::{FOREST_AUTHORITY_SEED, FOREST_SEED, MAX_TAG_LENGTH, NODE_SEED, TREE_SEED};
use crate::errors::DipErrors;
use crate::events;
use crate::state::{Forest, Node, Tree};

pub fn create_tree(ctx: Context<CreateTree>, title: String) -> Result<()> {
    msg!("Creating the tree");

    transfer(
        CpiContext::new(
            ctx.accounts.token_program.to_account_info(),
            Transfer {
                from: ctx.accounts.creator_account.to_account_info(),
                to: ctx.accounts.admin_account.to_account_info(),
                authority: ctx.accounts.signer.to_account_info(),
            },
        ),
        ctx.accounts.forest.tree_creation_fee,
    )?;

    let tree = &mut ctx.accounts.tree;
    tree.forest = ctx.accounts.forest.key();
    tree.root_node = ctx.accounts.root_node.key();
    tree.title = title.clone();

    let root_node = &mut ctx.accounts.root_node;
    root_node.tree = tree.key();
    root_node.parent = Pubkey::default();
    root_node.tags.push(title.clone());

    emit!(events::NewTree {
        forest: ctx.accounts.forest.key(),
        tree: ctx.accounts.tree.key(),
        title: title,
    });

    Ok(())
}

#[derive(Accounts)]
#[instruction(title: String)]
pub struct CreateTree<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,

    /// CHECK: Check against forest
    pub admin: AccountInfo<'info>,

    /// The account that manages tokens
    /// CHECK: Safe because this read-only account only gets used as a constraint
    #[account(
        seeds = [
            FOREST_AUTHORITY_SEED.as_bytes(),
            &forest.key().to_bytes()
        ],
        bump,
    )]
    pub forest_authority: UncheckedAccount<'info>,

    /// The forest
    #[account(
        seeds = [
            FOREST_SEED.as_bytes(),
            &forest.id.to_bytes(),
        ],
        bump,
        has_one = admin,
        has_one = vote_mint,
    )]
    pub forest: Account<'info, Forest>,

    /// Mint of the token used to pay the tree creation fee
    pub vote_mint: Account<'info, Mint>,

    #[account(
        init_if_needed,
        payer = signer,
        associated_token::mint = vote_mint,
        associated_token::authority = signer,
    )]
    pub creator_account: Box<Account<'info, TokenAccount>>,

    #[account(
        init_if_needed,
        payer = signer,
        associated_token::mint = vote_mint,
        associated_token::authority = admin,
    )]
    pub admin_account: Box<Account<'info, TokenAccount>>,

    /// The tree
    #[account(
        init,
        payer = signer,
        space = Tree::LEN,
        seeds = [
            TREE_SEED.as_bytes(),
            &forest.key().to_bytes(),
            &title.as_ref(),
        ],
        bump,
        constraint = title.len() <= MAX_TAG_LENGTH @ DipErrors::StringTooLong,
    )]
    pub tree: Box<Account<'info, Tree>>,

    /// The root node of the new tree
    #[account(
        init,
        payer = signer,
        space = Node::LEN,
        seeds = [
            NODE_SEED.as_bytes(),
            &tree.key().to_bytes(),
            Pubkey::default().to_bytes().as_ref(),
            &title.as_ref(),
        ],
        bump,
    )]
    pub root_node: Account<'info, Node>,

    /// Common Solana programs
    pub token_program: Program<'info, Token>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub system_program: Program<'info, System>,
    pub rent: Sysvar<'info, Rent>,
}
